### ISSUE TITLE
Overflow mongodb getStatistics

### ISSUE THREAD

**Issue by thiemol** [RocketChat/Rocket.Chat#2747]
Your Rocket.Chat version: 0.24

i get in the info panel an internal error. The log says:
Exception while invoking method 'getStatistics' MongoError: Plan executor error during find: Overflow sort stage buffered data usage of 33554604 bytes exceeds internal limit of 33554432 bytes

Is there a missing index by sorting? (http://askubuntu.com/questions/501937/how-to-increase-buffered-data-limit)

**Comment by engelgabriel** [comment]
Hi, we have just added some new indexes to try to fix this, but we are not sure what indexes were missing.

But, if your log says "Version: 0.22.0" it means that you are not running the latest build. 
What upgrade process did you do? How are you deploying Rocket.Chat?

**Comment by marceloschmidt** [comment]
If you can test with latest code from `develop`, we've added two indexes to the users collection.

**Comment by thiemol** [comment]
Sorry the Version mismatch was my fault in an updatescript. I am now on version 0.25 with the same issue. I'll try the develop Version next.
(own Rocket.Chat instance on a Ubuntu Linux machine without docker)

I20160405-08:43:41.417(2) (migrations.js:114) Migrations: Migrating from version 33 -> 37
I20160405-08:43:43.129(2) (migrations.js:114) Migrations: Finished migrating.
I20160405-08:43:54.081(2) ➔ |       Version: 0.25.0                                       | 
I20160405-08:43:54.081(2) ➔ |  Process Port: 3000                                         | 

I20160405-08:43:54.488(2) (percolate_synced-cron.js:71) SyncedCron: Scheduled "Generate and save statistics" next run @Tue Apr 05 2016 08:43:54 GMT+0200 (CEST)
I20160405-08:44:56.713(2) Exception while invoking method 'getStatistics' MongoError: Plan executor error during find: Overflow sort stage buffered data usage of 33554604 bytes exceeds internal limit of 33554432 bytes

**Comment by thiemol** [comment]
I20160419-14:14:23.167(2) (migrations.js:114) Migrations: Migrating from version 42 -> 43
I20160419-14:14:23.172(2) (migrations.js:114) Migrations: Running up() on version 43
I20160419-14:14:23.180(2) (migrations.js:114) Migrations: Finished migrating.

I20160419-14:14:35.428(2) ➔ |       Version: 0.27.0                                       | 

I20160419-14:14:58.226(2) Exception while invoking method 'getStatistics' MongoError: Plan executor error during find: Overflow sort stage buffered data usage of 33554604 bytes exceeds internal limit of 33554432 bytes     at Object.Future.wait

**Comment by marceloschmidt** [comment]
@thiemol how much RAM do you have?

**Comment by thiemol** [comment]
free 6.581MB

**Comment by sampaiodiego** [comment]
@thiemol how big is your instance? how many users, messages and rooms?

also, do you have access to the database? if so, can you please run the following script and post here the response? thanks:

```
db['rocketchat_message'].getIndexes();
db['rocketchat_room'].getIndexes();
db['rocketchat_subscription'].getIndexes();
db['users'].getIndexes();
```

**Comment by thiemol** [comment]
sure:

`> db['rocketchat_message'].getIndexes();
[
	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "rocketchat.rocketchat_message"
	},
	{
		"v" : 1,
		"key" : {
			"rid" : 1,
			"ts" : 1
		},
		"name" : "rid_1_ts_1",
		"ns" : "rocketchat.rocketchat_message"
	},
	{
		"v" : 1,
		"key" : {
			"editedAt" : 1
		},
		"name" : "editedAt_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"editedBy._id" : 1
		},
		"name" : "editedBy._id_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"rid" : 1,
			"t" : 1,
			"u._id" : 1
		},
		"name" : "rid_1_t_1_u._id_1",
		"ns" : "rocketchat.rocketchat_message"
	},
	{
		"v" : 1,
		"key" : {
			"expireAt" : 1
		},
		"name" : "expireAt_1",
		"ns" : "rocketchat.rocketchat_message",
		"expireAfterSeconds" : 0
	},
	{
		"v" : 1,
		"key" : {
			"_fts" : "text",
			"_ftsx" : 1
		},
		"name" : "msg_text",
		"ns" : "rocketchat.rocketchat_message",
		"weights" : {
			"msg" : 1
		},
		"default_language" : "english",
		"language_override" : "language",
		"textIndexVersion" : 2
	},
	{
		"v" : 1,
		"key" : {
			"pinnedBy._id" : 1
		},
		"name" : "pinnedBy._id_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"starred._id" : 1
		},
		"name" : "starred._id_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"file._id" : 1
		},
		"name" : "file._id_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"ts" : 1
		},
		"name" : "ts_1",
		"ns" : "rocketchat.rocketchat_message"
	},
	{
		"v" : 1,
		"key" : {
			"u._id" : 1
		},
		"name" : "u._id_1",
		"ns" : "rocketchat.rocketchat_message"
	},
	{
		"v" : 1,
		"key" : {
			"mentions.username" : 1
		},
		"name" : "mentions.username_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"pinned" : 1
		},
		"name" : "pinned_1",
		"ns" : "rocketchat.rocketchat_message",
		"sparse" : 1
	}
]
> db['rocketchat_room'].getIndexes();
[
	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "rocketchat.rocketchat_room"
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"name" : 1
		},
		"name" : "name_1",
		"ns" : "rocketchat.rocketchat_room",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"default" : 1
		},
		"name" : "default_1",
		"ns" : "rocketchat.rocketchat_room"
	},
	{
		"v" : 1,
		"key" : {
			"usernames" : 1
		},
		"name" : "usernames_1",
		"ns" : "rocketchat.rocketchat_room"
	},
	{
		"v" : 1,
		"key" : {
			"t" : 1
		},
		"name" : "t_1",
		"ns" : "rocketchat.rocketchat_room"
	},
	{
		"v" : 1,
		"key" : {
			"u._id" : 1
		},
		"name" : "u._id_1",
		"ns" : "rocketchat.rocketchat_room"
	}
]
> db['rocketchat_subscription'].getIndexes();
[
	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"rid" : 1,
			"u._id" : 1
		},
		"name" : "rid_1_u._id_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"u._id" : 1,
			"name" : 1,
			"t" : 1
		},
		"name" : "u._id_1_name_1_t_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"open" : 1
		},
		"name" : "open_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"alert" : 1
		},
		"name" : "alert_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"unread" : 1
		},
		"name" : "unread_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"ts" : 1
		},
		"name" : "ts_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"rid" : 1,
			"alert" : 1,
			"u._id" : 1
		},
		"name" : "rid_1_alert_1_u._id_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"rid" : 1,
			"roles" : 1
		},
		"name" : "rid_1_roles_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"ls" : 1
		},
		"name" : "ls_1",
		"ns" : "rocketchat.rocketchat_subscription"
	},
	{
		"v" : 1,
		"key" : {
			"desktopNotifications" : 1
		},
		"name" : "desktopNotifications_1",
		"ns" : "rocketchat.rocketchat_subscription",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"mobilePushNotifications" : 1
		},
		"name" : "mobilePushNotifications_1",
		"ns" : "rocketchat.rocketchat_subscription",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"emailNotifications" : 1
		},
		"name" : "emailNotifications_1",
		"ns" : "rocketchat.rocketchat_subscription",
		"sparse" : 1
	}
]
> db['users'].getIndexes();
[
	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "rocketchat.users"
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"username" : 1
		},
		"name" : "username_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"emails.address" : 1
		},
		"name" : "emails.address_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.resume.loginTokens.hashedToken" : 1
		},
		"name" : "services.resume.loginTokens.hashedToken_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.resume.loginTokens.token" : 1
		},
		"name" : "services.resume.loginTokens.token_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"services.resume.haveLoginTokensToDelete" : 1
		},
		"name" : "services.resume.haveLoginTokensToDelete_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"services.resume.loginTokens.when" : 1
		},
		"name" : "services.resume.loginTokens.when_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.facebook.id" : 1
		},
		"name" : "services.facebook.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.github.id" : 1
		},
		"name" : "services.github.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.google.id" : 1
		},
		"name" : "services.google.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.meteor-developer.id" : 1
		},
		"name" : "services.meteor-developer.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.email.verificationTokens.token" : 1
		},
		"name" : "services.email.verificationTokens.token_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.password.reset.token" : 1
		},
		"name" : "services.password.reset.token_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.twitter.id" : 1
		},
		"name" : "services.twitter.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.github_enterprise.id" : 1
		},
		"name" : "services.github_enterprise.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.gitlab.id" : 1
		},
		"name" : "services.gitlab.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.wordpress.id" : 1
		},
		"name" : "services.wordpress.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"services.linkedin.id" : 1
		},
		"name" : "services.linkedin.id_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"roles" : 1
		},
		"name" : "roles_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"name" : 1
		},
		"name" : "name_1",
		"ns" : "rocketchat.users"
	},
	{
		"v" : 1,
		"key" : {
			"lastLogin" : 1
		},
		"name" : "lastLogin_1",
		"ns" : "rocketchat.users"
	},
	{
		"v" : 1,
		"key" : {
			"status" : 1
		},
		"name" : "status_1",
		"ns" : "rocketchat.users"
	},
	{
		"v" : 1,
		"key" : {
			"active" : 1
		},
		"name" : "active_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"statusConnection" : 1
		},
		"name" : "statusConnection_1",
		"ns" : "rocketchat.users",
		"sparse" : 1
	},
	{
		"v" : 1,
		"key" : {
			"type" : 1
		},
		"name" : "type_1",
		"ns" : "rocketchat.users"
	}
]
> 
`

**Comment by sampaiodiego** [comment]
@thiemol ok, seems you have all indexes needed.
I'm still investigating what could be the problem.

how big is your instance? how many users, messages and rooms do you have?

our [demo server](https://demo.rocket.chat/) is one of the biggest servers I've known and we don't have such issue:
users: 58000+
rooms: 20000+
msgs: 450000+

**Comment by thiemol** [comment]
@sampaiodiego  as i said, it is a small instance :)
users: 17
rooms: 23
msgs: 1200+

All other functions are working perfect, so if i find the cause of the issue, i'll post it here.

**Comment by Sing-Li** [comment]
@thiemol  - thanks!

As @marceloschmidt  requested earlier - it will be helpful if you can tell us what machine or VPS configuration you are running on -   number of cores?  size of memory?  size of disk?  OS? 

Just "free mem available" is not meaningful.

**Comment by sampaiodiego** [comment]
oh, this error on a small instances is even worse .. :disappointed:

**Comment by thiemol** [comment]
@Sing-Li @marceloschmidt 
VPS: 3.13.0-042stab112.15 Ubuntu 14.04 minimal
RAM 8GB Free 7GB
Disk 286GB Free 281GB
Cores: 4

**Comment by Sing-Li** [comment]
@thiemol   thanks again.

Can you also let us know:

1) your mongo server version 
2) the CLI (command line) switches/flags you used when starting mongo
3) if you are running the mongod on the same VPS server
4) if you are running other app that uses mongo against the mongo db server

**Comment by thiemol** [comment]
@Sing-Li Thanks for the support
1) db version v3.0.11
2) export MONGO_URL=mongodb://localhost:27017/rocketchat
3) yes
4) no

**Comment by engelgabriel** [comment]
Is your MongoDB running in replica set mode?



-- 
Sent from Mobile

**Comment by Sing-Li** [comment]
@thiemol  No problem.  For item (2)

Please let us know how you startup your Mongo SERVER (not the client).  What command line switches and flags are you using to start-up the server?

A mongo server can take up to 1 GB or RAM by itself when starting up. Curious to see you have both RC and mongo server running and still have 7 GB out of  8GB free.

Thanks again.

**Comment by thiemol** [comment]
It is not runnung in replicat set mode.
I follow the tutorial, so i starts mongo with the default configuration.

The content of mongod.conf is
```
# mongod.conf

# for documentation of all options, see:
#   http://docs.mongodb.org/manual/reference/configuration-options/

# Where and how to store data.
storage:
  dbPath: /var/lib/mongodb
  journal:
    enabled: true
#  engine:
#  mmapv1:
#  wiredTiger:

# where to write logging data.
systemLog:
  destination: file
  logAppend: true
  path: /var/log/mongodb/mongod.log

# network interfaces
net:
  port: 27017
  bindIp: 127.0.0.1


#processManagement:

#security:

#operationProfiling:

#replication:

#sharding:

## Enterprise-Only Options:

#auditLog:

#snmp:

```

**Comment by thiemol** [comment]
Ok after installing a second instance and import the last backup i get the same issues. So i imported the collections one by one and the failure only appears, when i import the rocketchat_statistics collection. 
So i tested in the second instance to drop the collection and after the refresh the failure was gone. 
I did the same on the first instance and it worked as well. 

Solution: Drop and rebuild of rocketchat_statistics. 
Cause: ?

Thanks for the support!

**Comment by sampaiodiego** [comment]
@thiemol thanks for the information =)

because of it we found the problem.. I just fixed it in 567ea3a
so, it shouldn't be a problem (to anyone) any more.

thank you
